marlowe-playground-client/src/MarloweEditor/State.purs

```diff
@@ -0,0 +1,222 @@
+import Marlowe (SPParams_)
+import Marlowe as Server
+import Marlowe.Holes (fromTerm)
+import Marlowe.Linter as Linter
+import Marlowe.Monaco (updateAdditionalContext)
+import Marlowe.Parser (parseContract)
+import Marlowe.Semantics (Contract, emptyState)
+import Marlowe.Symbolic.Types.Request as MSReq
+import MarloweEditor.Types (Action(..), AnalysisState(..), State, _analysisState, _bottomPanelView, _editorErrors, _editorWarnings, _keybindings, _selectedHole, _showBottomPanel)
+import Monaco (IMarker, isError, isWarning)
+import Network.RemoteData (RemoteData(..))
+import Network.RemoteData as RemoteData
+import StaticAnalysis.Reachability (getUnreachableContracts, startReachabilityAnalysis)
+import Servant.PureScript.Ajax (AjaxError)
+import Servant.PureScript.Settings (SPSettings_)
+import StaticData (marloweBufferLocalStorageKey)
+import StaticData as StaticData
+import Text.Pretty (pretty)
+import Types (WebData)
+
+handleAction ::
+  forall m.
+  MonadAff m =>
+  SPSettings_ SPParams_ ->
+  Action ->
+  HalogenM State Action ChildSlots Void m Unit
+handleAction _ (ChangeKeyBindings bindings) = do
+  assign _keybindings bindings
+  void $ query _marloweEditorPageSlot unit (Monaco.SetKeyBindings bindings unit)
+
+handleAction _ (HandleEditorMessage (Monaco.TextChanged text)) = do
+  modify_
+    ( set _selectedHole Nothing
+        <<< set _hasUnsavedChanges' true
+    )
+  liftEffect $ LocalStorage.setItem marloweBufferLocalStorageKey text
+  analysisState <- use _analysisState
+  let
+    parsedContract = parseContract text
+
+    unreachableContracts = getUnreachableContracts analysisState
+
+    (Tuple markerData additionalContext) = Linter.markers unreachableContracts parsedContract
+  markers <- query _marloweEditorPageSlot unit (Monaco.SetModelMarkers markerData identity)
+  void $ traverse editorSetMarkers markers
+  {-
+    There are three different Monaco objects that require the linting information:
+      * Markers
+      * Code completion (type aheads)
+      * Code suggestions (Quick fixes)
+     To avoid having to recalculate the linting multiple times, we add aditional context to the providers
+     whenever the code changes.
+  -}
+  providers <- query _marloweEditorPageSlot unit (Monaco.GetObjects identity)
+  case providers of
+    Just { codeActionProvider: Just caProvider, completionItemProvider: Just ciProvider } -> pure $ updateAdditionalContext caProvider ciProvider additionalContext
+    _ -> pure unit
+
+handleAction _ (HandleDragEvent event) = liftEffect $ FileEvents.preventDefault event
+
+handleAction settings (HandleDropEvent event) = do
+  liftEffect $ FileEvents.preventDefault event
+  contents <- liftAff $ readFileFromDragEvent event
+  void $ editorSetValue contents
+  -- updateContractInState contents --FIXME
+  -- FIXME: check but I don't think this is needed after the split
+  -- setOraclePrice settings
+  -- pure unit is only here so the comment stays in place
```

Comment:
🙀

Reply:
I removed this but didnt pushed the changes 😅